[Enhancement] Support Named Parameters in Scala Functions

EdwardArchive opened the pull request:
> **Issue**: [#66227](https://github.com/StarRocks/starrocks/issues/66227)  
> **Related PR**: [#66206](https://github.com/StarRocks/starrocks/pull/66206)  
> **Status**: Developing on SSRF (Named Parameter done)  
> **Authors**: @EdwardArchive, with review from @alvin-celerdata
## Why I'm doing:

<html>
<body>
<!--StartFragment--><h1 dir="auto" style="box-sizing: border-box; font-size: 2em; margin-top: 0px !important; margin-right: 0px; margin-bottom: 16px; margin-left: 0px; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid rgba(209, 217, 224, 0.7); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Named Arguments Support for Scalar Functions</h1><h2 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid rgba(209, 217, 224, 0.7); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Summary</h2><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">This PR implements<span> </span><strong style="box-sizing: border-box; font-weight: 600;">Named Arguments</strong><span> </span>support for scalar functions in StarRocks, using<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: rgba(129, 139, 152, 0.12); border-radius: 6px;">http_request()</code><span> </span>as the first function to leverage this feature.</p><h2 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid rgba(209, 217, 224, 0.7); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Features</h2><ul dir="auto" style="box-sizing: border-box; padding: 0px; margin-top: 0px; margin-bottom: 16px; position: relative; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">Named Arguments syntax</strong>:<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: rgba(129, 139, 152, 0.12); border-radius: 6px;">function(param =&gt; value, ...)</code></li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">Default values</strong>: Optional parameters can be omitted</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">Positional calls</strong>: Traditional positional syntax still works</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">User-friendly error messages</strong>: Clear hints for common mistakes</li></ul><h2 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid rgba(209, 217, 224, 0.7); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Usage Examples</h2><div class="highlight highlight-source-sql notranslate position-relative overflow-auto" dir="auto" style="box-sizing: border-box; position: relative !important; overflow: visible !important; margin-bottom: 16px; background-color: rgb(255, 255, 255); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><pre class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; margin-top: 0px; margin-bottom: 0px; tab-size: 4; overflow-wrap: normal; padding: 16px; overflow: auto; line-height: 1.45; color: rgb(31, 35, 40); background-color: rgb(246, 248, 250); border-radius: 6px; word-break: normal; min-height: 52px;"><span class="pl-c" style="box-sizing: border-box; color: rgb(89, 99, 110);"><span class="pl-c" style="box-sizing: border-box; color: rgb(89, 99, 110);">--</span> Named Arguments (any order, omit optional params)</span>
<span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">SELECT</span> http_request(url <span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">=&gt;</span> <span class="pl-s" style="box-sizing: border-box; color: rgb(10, 48, 105);"><span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span>https://api.example.com<span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span></span>);
<span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">SELECT</span> http_request(url <span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">=&gt;</span> <span class="pl-s" style="box-sizing: border-box; color: rgb(10, 48, 105);"><span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span>https://api.example.com<span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span></span>, method <span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">=&gt;</span> <span class="pl-s" style="box-sizing: border-box; color: rgb(10, 48, 105);"><span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span>POST<span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span></span>, body <span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">=&gt;</span> <span class="pl-s" style="box-sizing: border-box; color: rgb(10, 48, 105);"><span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span>{}<span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span></span>);
<span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">SELECT</span> http_request(method <span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">=&gt;</span> <span class="pl-s" style="box-sizing: border-box; color: rgb(10, 48, 105);"><span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span>POST<span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span></span>, url <span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">=&gt;</span> <span class="pl-s" style="box-sizing: border-box; color: rgb(10, 48, 105);"><span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span>https://api.example.com<span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span></span>);  <span class="pl-c" style="box-sizing: border-box; color: rgb(89, 99, 110);"><span class="pl-c" style="box-sizing: border-box; color: rgb(89, 99, 110);">--</span> order doesn't matter</span>

<span class="pl-c" style="box-sizing: border-box; color: rgb(89, 99, 110);"><span class="pl-c" style="box-sizing: border-box; color: rgb(89, 99, 110);">--</span> Positional Arguments (still supported)</span>
<span class="pl-k" style="box-sizing: border-box; color: rgb(207, 34, 46);">SELECT</span> http_request(<span class="pl-s" style="box-sizing: border-box; color: rgb(10, 48, 105);"><span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span>https://api.example.com<span class="pl-pds" style="box-sizing: border-box; color: rgb(10, 48, 105);">'</span></span>);</pre><div class="zeroclipboard-container position-absolute right-0 top-0" style="box-sizing: border-box; position: absolute !important; top: 0px !important; right: 0px !important;"><clipboard-copy aria-label="Copy" class="ClipboardButton btn js-clipboard-copy m-2 p-0" data-copy-feedback="Copied!" data-tooltip-direction="w" value="-- Named Arguments (any order, omit optional params)
SELECT http_request(url =&gt; 'https://api.example.com');
SELECT http_request(url =&gt; 'https://api.example.com', method =&gt; 'POST', body =&gt; '{}');
SELECT http_request(method =&gt; 'POST', url =&gt; 'https://api.example.com');  -- order doesn't matter

-- Positional Arguments (still supported)
SELECT http_request('https://api.example.com');" tabindex="0" role="button" style="box-sizing: border-box; position: relative; display: inline-block; padding: 0px !important; font-size: 14px; font-weight: 500; line-height: 20px; white-space: nowrap; vertical-align: middle; cursor: pointer; user-select: none; border: 1px solid rgb(209, 217, 224); border-radius: 6px; appearance: none; color: rgb(37, 41, 46); background-color: rgb(246, 248, 250); box-shadow: none; transition: color 80ms cubic-bezier(0.33, 1, 0.68, 1), background-color, box-shadow, border-color; margin: 8px !important;"><svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true" class="octicon octicon-copy js-clipboard-copy-icon m-2"></svg></clipboard-copy></div></div><h2 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid rgba(209, 217, 224, 0.7); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Error Messages</h2><markdown-accessiblity-table data-catalyst="" style="box-sizing: border-box; display: block; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">
Scenario | Error Message
-- | --
Missing required param | http_request() required parameter 'url' is missing
Unknown param (with hint) | http_request() unknown parameter 'URL'. Did you mean 'url'?
Duplicate param | http_request() duplicate parameter 'url'
NULL for required param | http_request() required parameter 'url' cannot be NULL
No arguments | http_request() requires at least 1 argument(s). Missing required parameter(s): 'url'

</markdown-accessiblity-table><h3 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: 600; line-height: 1.25; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Architecture Flow</h3><div class="snippet-clipboard-content notranslate position-relative overflow-auto" style="box-sizing: border-box; position: relative !important; overflow: auto !important; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><pre class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; margin-top: 0px; margin-bottom: 16px; tab-size: 4; overflow-wrap: normal; padding: 16px; overflow: auto; line-height: 1.45; color: rgb(31, 35, 40); background-color: rgb(246, 248, 250); border-radius: 6px;"><code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0px; margin: 0px; white-space: pre; background: rgba(0, 0, 0, 0); border-radius: 6px; word-break: normal; border: 0px; display: inline; overflow: visible; line-height: inherit; overflow-wrap: normal;">SQL: http_request(url =&gt; '...', method =&gt; 'POST')
         │
         ▼
┌─────────────────────────────────────┐
│ 1. Parser (StarRocks.g4)            │
│    Parse `param =&gt; value` syntax    │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 2. AstBuilder.java                  │
│    Create NamedArgument AST nodes   │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 3. FunctionParams.java              │
│    Separate exprs[] + exprsNames[]  │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 4. ExpressionAnalyzer.java          │
│    Branch: Named vs Positional      │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 5. FunctionAnalyzer.java            │
│    Validate &amp; lookup function       │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 6. FunctionParams.java              │
│    Reorder args &amp; append defaults   │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 7. Backend (C++)                    │
│    Receive all 8 columns            │
└─────────────────────────────────────┘
</code></pre><div class="zeroclipboard-container position-absolute right-0 top-0" style="box-sizing: border-box; position: absolute !important; top: 0px !important; right: 0px !important;"><clipboard-copy aria-label="Copy" class="ClipboardButton btn js-clipboard-copy m-2 p-0" data-copy-feedback="Copied!" data-tooltip-direction="w" value="SQL: http_request(url =&gt; '...', method =&gt; 'POST')
         │
         ▼
┌─────────────────────────────────────┐
│ 1. Parser (StarRocks.g4)            │
│    Parse `param =&gt; value` syntax    │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 2. AstBuilder.java                  │
│    Create NamedArgument AST nodes   │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 3. FunctionParams.java              │
│    Separate exprs[] + exprsNames[]  │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 4. ExpressionAnalyzer.java          │
│    Branch: Named vs Positional      │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 5. FunctionAnalyzer.java            │
│    Validate &amp; lookup function       │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 6. FunctionParams.java              │
│    Reorder args &amp; append defaults   │
└─────────────────────────────────────┘
         │
         ▼
┌─────────────────────────────────────┐
│ 7. Backend (C++)                    │
│    Receive all 8 columns            │
└─────────────────────────────────────┘" tabindex="0" role="button" style="box-sizing: border-box; position: relative; display: inline-block; padding: 0px !important; font-size: 14px; font-weight: 500; line-height: 20px; white-space: nowrap; vertical-align: middle; cursor: pointer; user-select: none; border: 1px solid rgb(209, 217, 224); border-radius: 6px; appearance: none; color: rgb(37, 41, 46); background-color: rgb(246, 248, 250); box-shadow: none; transition: color 80ms cubic-bezier(0.33, 1, 0.68, 1), background-color, box-shadow, border-color; margin: 8px !important;"><svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true" class="octicon octicon-copy js-clipboard-copy-icon m-2"></svg></clipboard-copy></div></div><h2 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid rgba(209, 217, 224, 0.7); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Backward Compatibility</h2><ul dir="auto" style="box-sizing: border-box; padding: 0px; margin-top: 0px; margin-bottom: 16px; position: relative; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">No impact on existing functions</strong>: Only functions with<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: rgba(129, 139, 152, 0.12); border-radius: 6px;">named_args</code><span> </span>defined use this path</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">Positional calls still work</strong>:<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: rgba(129, 139, 152, 0.12); border-radius: 6px;">http_request('url')</code><span> </span>works as expected</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">Varargs functions excluded</strong>: Functions like<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: rgba(129, 139, 152, 0.12); border-radius: 6px;">concat()</code><span> </span>are not affected</li></ul><h2 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid rgba(209, 217, 224, 0.7); color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Constraints</h2><ul dir="auto" style="box-sizing: border-box; padding: 0px; margin-top: 0px; margin-bottom: 0px !important; position: relative; color: rgb(31, 35, 40); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">No mixed mode</strong>: All arguments must be either named or positional (not both)</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">Case-sensitive</strong>: Parameter names are case-sensitive (<code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: rgba(129, 139, 152, 0.12); border-radius: 6px;">url</code><span> </span>≠<span> </span><code class="notranslate" style="box-sizing: border-box; font-family: &quot;Monaspace Neon&quot;, ui-monospace, SFMono-Regular, &quot;SF Mono&quot;, Menlo, Consolas, &quot;Liberation Mono&quot;, monospace; font-size: 11.9px; tab-size: 4; padding: 0.2em 0.4em; margin: 0px; white-space: break-spaces; background-color: rgba(129, 139, 152, 0.12); border-radius: 6px;">URL</code>)</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 24px;"><strong style="box-sizing: border-box; font-weight: 600;">Varargs not supported</strong>: Functions with variable arguments cannot use Named Arguments</li></ul><!--EndFragment-->
</body>
</html>




<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces first-class named arguments for scalar functions with robust validation, reordering, and defaults handling.
> 
> - Parser/AST: Extends grammar to accept `argumentItem` and `namedArgsFunctionCall`; `AstBuilder` constructs `NamedArgument` nodes and normalizes argument lists.
> - Analyzer: `ExpressionAnalyzer` detects named vs positional calls, validates structure, reorders to function signature, appends defaults, re-analyzes children, and enforces NULL constraints; supports positional calls against named-arg functions by filling defaults; improved error messaging.
> - Function layer: `FunctionAnalyzer` adds helpers to lookup with named args, validate duplicates/unknown/missing params, check NULLs post-reorder, reorder/apply defaults, and provide friendly errors; adjusts arg-count/type checks to accommodate named args.
> - Parser model updates: `FunctionParams` gains named-arg helpers/setters; `NamedArgument` carries position info and correct cloning.
> - Codegen: `gensrc/script/gen_functions.py` emits FE metadata for named args (arg names and Java default Exprs) into builtins.
> - Tests: New unit tests for validation and params behavior, plus analyzer tests for syntax/mixing/positional paths.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9939b11d22a7df08423f65ecf4dc1bd64b3ec483. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->






This could lead to runtime errors or unexpected behavior.

## What I'm doing:

This PR adds comprehensive validation for named arguments in scalar functions:
1. Validates that all required parameters (without defaults) are provided
2. Checks for duplicate parameter names
3. Validates that required parameters cannot be NULL
4. Provides user-friendly error messages with suggestions for typos
5. Moves validation logic from `FunctionParams` to `FunctionAnalyzer` to resolve module dependency issues

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

## If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.1 
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

EdwardArchive committed:
Split over Named paramter function

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

alvin-celerdata commented:
@EdwardArchive, one thing: could you add some SQL tests to verify that your changes work?

EdwardArchive commented:
> @EdwardArchive, one thing: could you add some SQL tests to verify that your changes work?
Hi @alvin-celerdata, do you have any ideas for creating test SQL on this PR? This PR doesn't have any Scala function changes, so should the test SQL be for running non-named parameter functions?

EdwardArchive committed:
Add unit tests for named argument validation in FunctionParams

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive commented:
> > > @EdwardArchive, one thing: could you add some SQL tests to verify that your changes work?
> > > Hi @alvin-celerdata, do you have any ideas for creating test SQL on this PR? This PR doesn't have any Scala function changes, so should the test SQL be for running non-named parameter functions?
> 
> After suggesting, I found out that there are no functions that have named argument functions, right now. It makes it difficult to add SQL test, but I believe UT should be added to cover the changes.

Hi I just added.

alvin-celerdata commented:
@cursor review

EdwardArchive commented:
@alvin-celerdata 
Hi, is there any comment on my code? thanks

EdwardArchive committed:
[Enhancement] Add NULL validation for required Named Arguments parameters
  - Required parameters (no default) cannot be NULL
  - Move FunctionParams methods to FunctionAnalyzer (fix module dependency)
  - Add NULL validation tests

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive committed:
fix : FunctionAnalyzer to use getFunctionName and update tests for consistency

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

alvin-celerdata commented:
@cursor review

EdwardArchive committed:
BUG FIX: Validate structure BEFORE reordering

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive committed:
Fix: Explicitly return null to trigger proper 'No matching function' error

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive commented:
@alvin-celerdata 
Hi, this PR(https://github.com/StarRocks/starrocks/pull/66206) depends on the Named Arguments feature I'm working on in another branch. I'd like to set that
branch as the base for this PR (stacked PRs), but I can only select branches from the upstream repo, not from
my fork. Is there a recommended way to handle dependent PRs here?

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@mergify rebase

EdwardArchive committed:
[Enhancement] Add mixed args validation and fix string escape in gen_functions

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

alvin-celerdata commented:
@cursor review

EdwardArchive committed:
[Enhancement] Implement validation for mixing named and positional arguments in function calls

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive committed:
Fix : Except '=' expression on namedArgument

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive committed:
[Enhancement] Refactor function call parsing to separate named and positional arguments

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive committed:
[Fix] Required parameter validation assumes positional ordering

stephen-shelby commented on fe/fe-grammar/src/main/antlr/com/starrocks/grammar/StarRocks.g4:
```diff
@@ -2482,7 +2486,6 @@ namedArgumentList
 
 namedArgument
     : identifier '=>' expression                                                        #namedArguments
-    | identifier '=' expression                                                         #namedArguments
```
this change will break existing queries.

EdwardArchive committed:
Fix : revoke Except '=' expression on namedArgument

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

EdwardArchive committed:
fix : ANTLR grammar error

alvin-celerdata commented:
@cursor review

EdwardArchive committed:
Fix : revert ANTLR grammar error

Signed-off-by: Edward Kim <kjjs1996@gmail.com>

stdpain commented:
postgresql use '=>' or ':=' https://www.postgresql.org/docs/current/sql-syntax-calling-funcs.html

EdwardArchive committed:
[Enhancement] Separate functionNamedArgument (=> only) from namedArgument for scalar functions

Resolves grammar ambiguity where count(a=b) could be parsed as named argument.
Scalar functions now only accept => syntax, while CALL procedures keep both = and =>.

Signed-off-by: EdwardKim <kjjs1996@gmail.com>

EdwardArchive committed:
[Refactor] Clean up dead code and restore argumentList grammar structure
- Remove dead namedArgumentList rule (orphaned after argumentItem refactoring)
- Restore argumentList to original structure (namedArgumentList | expressionList)
  to block mixed args at grammar level
- Remove visitArgumentItem (no longer needed)
- Remove unreachable mixed-args checks in FunctionAnalyzer and ExpressionAnalyzer
  (grammar separation already prevents mixing)
- Extract buildNamedArgument() helper to deduplicate visitNamedArguments
  and visitFunctionNamedArgument
- Remove unreachable testMixingNamedAndPositionalDetection test

Signed-off-by: EdwardKim <kjjs1996@gmail.com>
